### PR TITLE
(Paid code) Adds in battle royale mode (this time with a bit of theming)

### PR DESCRIPTION
**Pull request by Kmc2000**
### Intent of your Pull Request

Council member YogurtShrimp69 requested this as a commission.

Latejoins are disabled and all crew are put into a battle bus!
https://streamable.com/ba98i
https://i.gyazo.com/b53f366390d5b8b2aa527f0453c6025a.mp4

The storm closes in on you every 4 minutes (Credits to nichlas for that code)

You can find chests randomly littered in areas, these range from Common to LEGENDARY and have according items and medical stuff.

This is designed as an adminbus mode ONLY!
If you force this on anything other than box the order of zones will be WEIRD.

The winner takes the epic W

If you leave the station you lose your chance at the DUB, same goes for dying, duh.

The winner is checked every 30 seconds to avoid massive lag.

This will need stress testing with lots of players!

https://www.youtube.com/watch?v=kVRQ4VT-so8

#### Changelog

:cl: Kmc2000
rscadd: Added battle royale mode as an adminbus roundtype 
/:cl:

### CODE REVIEW

**Comment by ghost** [comment]
L-MRP

**Comment by Kmc2000** [comment]
You can easily rp in this mode
Rp as an insane person who kills all the crew!
Every brazilian newfriend's dream

I make dreams happen, Floyd. What do YOU do?

**Comment by ghost** [comment]
you dont thank the bus driver when u get out so it sucks

**Comment by Kmc2000** [comment]
The bus driver is a subsystem
Machines don't need thanks

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+		SSticker.mode.check_finished(TRUE)
+		SSticker.force_ending = 1
+		to_chat(world, "<span_class='ratvar'>L. Nobody wins!</span>")
+		SEND_SOUND(world, 'yogstation/sound/effects/battleroyale/L.ogg')
+		return
+	if(royalers.len == 1) //We have a wiener!
+		SSticker.mode.check_finished(TRUE)
+		SSticker.force_ending = 1
+		var/mob/living/dub = pick(royalers)
+		to_chat(world, "<img src='https://cdn.discordapp.com/attachments/351367327184584704/539903688857092106/victoryroyale.png'>")
```
Comment (babydoll) [inline yogstation/code/game/gamemodes/battle_royale/battleroyale.dm:100]: ❓ ❔ ❓

Comment (Kmc2000) [reply yogstation/code/game/gamemodes/battle_royale/battleroyale.dm:100]: Victory royale.png

**Review by alexkar598 — CHANGES_REQUESTED** [review]
lack of default dance

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+	for(var/i = 0, i < fortnite_gamers, ++i) //Iterate for number of lobby-ers
+		var/mob/dead/new_player/newvirgin = pick(virgins)
+		if(isobserver(newvirgin) || !newvirgin.mind || !newvirgin.client)
+			virgins -= newvirgin //No ghosts or AFKS please
+			continue
+		var/datum/mind/virgin = newvirgin.mind
+		queued += virgin
+		virgin.assigned_role = ROLE_BATTLEROYALE
+		virgin.special_role = ROLE_BATTLEROYALE
+		log_game("[key_name(virgin)] has been forcibly conscripted into battle royale.")
```
Comment (monster860) [inline yogstation/code/game/gamemodes/battle_royale/battleroyale.dm]: This code makes me want to vomit.

**Review by monster860 — CHANGES_REQUESTED** [review]

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm (hunk 2 above)
Comment (Kmc2000) [reply yogstation/code/game/gamemodes/battle_royale/battleroyale.dm]: What's wrong with it?
:b1:

**Comment by ThatLing** [comment]
```
Latejoins are disabled
```
pr closed

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+	required_players = 1 //Everyone is an antag in this mode
+	required_enemies = 1
+	recommended_enemies = 1
+	antag_flag = ROLE_BATTLEROYALE
+	enemy_minimum_age = 0
+
+	announce_span = "warning"
+	announce_text = "Attention ALL space station 13 crewmembers,\n\
+	<span class='danger'><b>John wick</b></span> is in grave danger and he NEEDS your help to wipe all the squads in the tilted towers. To do this, he'll need\n\
+	<span class='notice'><i>your credit card number, the three numbers on the back and the expiration month <b> AND </b> year</i></span>. But you gotta be <b>quick</b>, so that John can secure the bag and achieve the epic victory ROYAL.\n\
```
Comment (1fbff5f83b23d39d38b1dfcb4cac8d9b) [inline yogstation/code/game/gamemodes/battle_royale/battleroyale.dm:17]: ROYAL -> ROYALE

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+	<i>Be the last man standing at the end of the game to win</i>"
```
Comment (1fbff5f83b23d39d38b1dfcb4cac8d9b) [inline yogstation/code/game/gamemodes/battle_royale/battleroyale.dm]: period

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+	false_report_weight = 0
+	required_players = 1 //Everyone is an antag in this mode
+	required_enemies = 1
+	recommended_enemies = 1
+	antag_flag = ROLE_BATTLEROYALE
+	enemy_minimum_age = 0
+
+	announce_span = "warning"
+	announce_text = "Attention ALL space station 13 crewmembers,\n\
+	<span class='danger'><b>John wick</b></span> is in grave danger and he NEEDS your help to wipe all the squads in the tilted towers. To do this, he'll need\n\
```
Comment (1fbff5f83b23d39d38b1dfcb4cac8d9b) [inline yogstation/code/game/gamemodes/battle_royale/battleroyale.dm]: perhaps change fortnite specific things to ss13 stuff, e.g. change john wick to george melons

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+		SEND_SOUND(virgin.current, 'yogstation/sound/effects/battleroyale/battlebus.ogg')
+		virgin.add_antag_datum(antag_datum_type)
+		if(!GLOB.thebattlebus) //Ruhoh.
+			virgin.current.forceMove(pick(GLOB.start_landmarks_list))
+			message_admins("There is no battle bus! Attempting to spawn players at random.")
+			continue
+		virgin.current.forceMove(GLOB.thebattlebus)
+		virgin.current.add_trait(TRAIT_XRAY_VISION) //so they can see where theyre dropping
+		virgin.current.update_sight()
+		to_chat(virgin.current, "<font_color='red'><b> You are now in the battle bus! Click it to exit </b></font>")
```
Comment (1fbff5f83b23d39d38b1dfcb4cac8d9b) [inline yogstation/code/game/gamemodes/battle_royale/battleroyale.dm]: period

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+		SSticker.mode.check_finished(TRUE)
+		SSticker.force_ending = 1
+		to_chat(world, "<span_class='ratvar'>L. Nobody wins!</span>")
+		SEND_SOUND(world, 'yogstation/sound/effects/battleroyale/L.ogg')
+		return
+	if(royalers.len == 1) //We have a wiener!
+		SSticker.mode.check_finished(TRUE)
+		SSticker.force_ending = 1
+		var/mob/living/dub = pick(royalers)
+		to_chat(world, "<img src='https://cdn.discordapp.com/attachments/351367327184584704/539903688857092106/victoryroyale.png'>")
```
Comment (1fbff5f83b23d39d38b1dfcb4cac8d9b) [inline yogstation/code/game/gamemodes/battle_royale/battleroyale.dm:100]: Discord might delete the image, you should look into keeping it on the server

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+	var/datum/objective/O = new /datum/objective/survive() //SURVIVE.
+	O.owner = owner
+	objectives += O
+	var/mob/living/carbon/human/tfue = owner.current
+	tfue.equipOutfit(/datum/outfit/battleroyale, visualsOnly = FALSE)
+
+/datum/antagonist/battleroyale/greet()
+	SEND_SOUND(owner.current, 'yogstation/sound/effects/battleroyale/greet_br.ogg')
+	to_chat(owner.current, "<span_class='bigbold'>Welcome contestant!</span>")
+	to_chat(owner.current, "<span_class='danger'>You have been entered into nanotrasen's up and coming TV show! : <b> LAST MAN STANDING </b>. \n\ KILL YOUR COWORKERS TO ACHIEVE THE VICTORY ROYALE!. Attempting to leave the station will disqualify you from the round!</span>")
```
Comment (1fbff5f83b23d39d38b1dfcb4cac8d9b) [inline yogstation/code/game/gamemodes/battle_royale/battleroyale.dm]: nanotrasen -> Nanotrasen
TV show! : -> TV show:
ROYALE!. -> ROYALE!

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+	shoes = /obj/item/clothing/shoes/jackboots
+	ears = /obj/item/radio/headset
+	r_pocket = /obj/item/bikehorn
+	l_pocket = /obj/item/crowbar
+	back = /obj/item/storage/backpack
+	id = /obj/item/card/id/syndicate
+
+/obj/structure/battle_bus
+	name = "The battle bus"
+	desc = "Quit screwin' around!. Can we get some of that jumping music?. Click it to exit!"
```
Comment (1fbff5f83b23d39d38b1dfcb4cac8d9b) [inline yogstation/code/game/gamemodes/battle_royale/battleroyale.dm]: Exclamation mark period, question mark period. Where did you learn doing that?

File: yogstation/code/game/gamemodes/battle_royale/storm.dm [hunk 10]
```diff
@@ -0,0 +1,72 @@
+/datum/weather/royale
+	telegraph_message = "<span class='userdanger'><i>The storm is closing, get away from the edge!</i></span>"
```
Comment (1fbff5f83b23d39d38b1dfcb4cac8d9b) [inline yogstation/code/game/gamemodes/battle_royale/storm.dm]: closing in? make sure to change all the other messages too

**Review by 1fbff5f83b23d39d38b1dfcb4cac8d9b — CHANGES_REQUESTED** [review]
Grammer nitpicks and suggestions

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm (hunk 3 above)
Comment (monster860) [reply yogstation/code/game/gamemodes/battle_royale/battleroyale.dm:17]: I think this was intentional.

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm (hunk 7 above)
Comment (Kmc2000) [reply yogstation/code/game/gamemodes/battle_royale/battleroyale.dm:100]: I tried that but cannot into HTML

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm (hunk 3 above)
Comment (Kmc2000) [reply yogstation/code/game/gamemodes/battle_royale/battleroyale.dm:17]: intended

**Review by 1fbff5f83b23d39d38b1dfcb4cac8d9b — APPROVED** [review]

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+			continue //No AFKS allowed!!!
+		if(player.onCentCom())
+			GLOB.battleroyale_players -= player
+			to_chat(player, "You left the station Z-level! You have been disqualified from battle royale.")
+			continue
+		else if(player.onSyndieBase())
+			GLOB.battleroyale_players -= player
+			to_chat(player, "You left the station Z-level! You have been disqualified from battle royale.")
+			continue
+		if(player.z != station_z)
```
Comment (Amelia0010) [inline yogstation/code/game/gamemodes/battle_royale/battleroyale.dm]:
```suggestion
		if(!is_station_level(player.z))
```

don't ignore multiz memes

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+		if(!player.client)
+			GLOB.battleroyale_players -= player
+			continue //No AFKS allowed!!!
+		if(player.onCentCom())
+			GLOB.battleroyale_players -= player
+			to_chat(player, "You left the station Z-level! You have been disqualified from battle royale.")
+			continue
+		else if(player.onSyndieBase())
+			GLOB.battleroyale_players -= player
+			to_chat(player, "You left the station Z-level! You have been disqualified from battle royale.")
```
Comment (Amelia0010) [inline yogstation/code/game/gamemodes/battle_royale/battleroyale.dm]: remove the z-level thing, just call it the station.

**Review by Amelia0010 — CHANGES_REQUESTED** [review]
we can run a final test after this, then merge if it's fine

**Review by Amelia0010 — APPROVED** [review]

File: yogstation/code/game/gamemodes/battle_royale/battleroyale.dm (hunk 1 above)
Comment (Altoids1) [reply yogstation/code/game/gamemodes/battle_royale/battleroyale.dm:100]: Does Discord actually permanently store all images?

Comment (ondrej008) [reply yogstation/code/game/gamemodes/battle_royale/battleroyale.dm:100]: if the message is not deleted, or the channel isn't deleted, or the server isn't deleted, then yes

Comment (Amelia0010) [reply yogstation/code/game/gamemodes/battle_royale/battleroyale.dm:100]: What is it with people necroing year-old PRs

**Pull request merged**